// react-navigation-bottom-tabs/lib/bottom-navigation.js
import React from 'react';
import { View, Text, Platform } from 'react-native';
import { createMaterialBottomTabNavigator } from 'react-navigation-material-bottom-tabs';
import Icon from 'react-native-vector-icons/MaterialIcons';
import { createBottomTabNavigator } from 'react-navigation-tabs';
function tabIcon(props) {
    const { badgeCount, name, ...rest } = props;
    if (name) {
        return ({ tintColor }) => (<View style={{ width: 24, height: 24, margin: !!badgeCount ? 5 : 0 }}>
				<Icon {...rest} name={name} style={{ color: tintColor, fontSize: 24, textAlign: 'center' }}/>
				{!!badgeCount && (<View style={{
            position: 'absolute',
            right: -6,
            top: -3,
            backgroundColor: 'red',
            borderRadius: 9,
            width: 18,
            height: 18,
            justifyContent: 'center',
            alignItems: 'center'
        }}>
						<Text style={{ color: 'white', fontSize: 12, fontWeight: 'bold' }}>
							{badgeCount}
						</Text>
					</View>)}
			</View>);
    }
    return null;
}
export function createBottomTab(screens, navigatorConfig) {
    const routeConfig = {};
    for (const [key, value] of Object.entries(screens)) {
        const { screen, icon, title } = value;
        routeConfig[key] = {
            screen: screen,
            navigationOptions: {
                title,
                tabBarIcon: tabIcon({ name: icon })
            }
        };
    }
    if (Platform.OS === 'ios') {
        return createBottomTabNavigator(routeConfig, navigatorConfig);
    }
    else {
        return createMaterialBottomTabNavigator(routeConfig, navigatorConfig);
    }
}
